working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+# Shared Memory Multithreading for Dart
+
+This proposal tries to address two long standing gaps which separate Dart from other more low-level languages:
+
+* Dart developers should be able to utilize multicore capabilities without hitting limitations of the isolate model.
+
+* Interoperability between Dart and native platforms (C/C++, Objective-C/Swift, Java/Kotlin) requires aligning concurrency models between Dart world and native world. The misalignment is not an issue when invoking simple native synchronous APIs from Dart, but it becomes a blocker when:
+
+  * working with APIs pinned to a specific thread (e.g. `@MainActor` APIs in Swift or UI thread APIs on Android) 
+
+  * dealing with APIs which want to call _back_ from native into Dart on an arbitrary thread.
+
+  * using Dart as a container for shared business logic - which can be invoked on an arbitrary thread by surrounding native application. 
+
+  Native code does not _understand_ the concept of isolates. 
+
+The biggest challenge with introducing shared memory multithreading into Dart is avoiding breaking existing code, which is written under the assumption of single-threaded access. We would like to avoid just outright breaking existing code in subtle ways. Consider for a moment a library which is maintains a global cache internally:  
+
+```     dart
+library foo;
+
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId;
+final _cache = <int, C>{};
+
+C makeObject() {
+  map[nextId] = C(nextId);
+  nextId++;
+  return map[nextId];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the same code becomes thread _unsafe_ in the presence of shared memory multithreading.
+
+## Shareable Data
+
+We propose to extend Dart with a shared memory multithreading but provide a clear type-based delineation between two concurrency worlds. **Type of an instance should provide a clear indication whether its fields can be concurrently mutated by another thread.** We are going to call such types _shareable_ and assume that the language provides some way to opt-in into this for classes:
+
+```  dart
+/// Mark class sharable by implementing a [Shareable]
+/// interface.
+class S implements Shareable {
+  // ...
+} 
+```
+
+To maintain heap isolation we restrict the kind of data that a shareable class can contain: **Fields of sharable classes can only contain references to instances of sharable classes.** There is however no requirement of immutability imposed on shareable classes.
+
+References to shareable instances can be passed between isolates within an isolate group directly: for example sending a shareable instance through `SendPort` will not copy it.
+
+```dart  
+class S implements Shareable {
+  int v = 0;
```

Comment:
How do concurrent reads and writes to these fields work? Are these fields effectively atomic? Do we make any guarantees about memory ordering? If not, is the best practice for developers to always use the atomics/locks mentioned below?

From the other comments it sounds like we haven't settled on a memory model yet. That's fine, but can you mention that here? Are there any aspects of the memory model that are settled?

Reply:
If the concurrency model is anything like people use it in Java, then no. You would write a class that uses something such as mutexes internally to guard access to multiple fields at the time for efficiency. Making every field atomic would make it rather slow. (Also that would work if you for example implement a list and keep an int for the length around, because one needs to atomically update two fields, not one.)

Reply:
Ah ok. So shareable means nothing other than "is allowed to be shared between isolates". It doesn't have any effect on the memory model etc.

Reply:
Actually, I guess shareable also means that the developer is making a strong claim that they're doing all the correct locking etc to make sure the class is thread safe.

Reply:
Added clarifications around memory model.